torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -175,6 +177,488 @@ std::vector<kir::Expr*> generateConditionalFromPredicate(
+//! without bounds-checking predicates.
+//! TODO: Merge this with PredicateElimination
+class IterationDomainAnalysis : private OptOutDispatch {
+ public:
+  //! Return true if the expression defining tv can be safely run
+  //! without a predicate
+  static bool canOmitPredicate(const TensorDomain* td) {
+    const auto gpu_lower = GpuLower::current();
+    for (size_t i = 0; i < td->nDims(); ++i) {
+      IterDomain* id = gpu_lower->caLoopMap().getConcreteMappedID(td->axis(i));
+      IterationDomainAnalysis id_analysis(id->fusion());
+      auto extent = id->extent();
+      id_analysis.handle(extent);
+      if (!id_analysis.isExact(extent)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  IterationDomainAnalysis(Fusion* fusion) : fusion_(fusion) {}
+
+  using OptOutDispatch::handle;
+
+  //! Check if val has nothing that prevents a loop using val as its
+  //! extent to omit a bounds-checking predicate
+  bool isExact(const Val* val) {
+    return exact_vals_.find(val) != exact_vals_.end();
+  }
+
+  //! Record val does not need a predicate.
+  void setExact(const Val* val) {
+    exact_vals_.insert(val);
+  }
+
+  void handle(Val* val) override {
+    if (val->definition() != nullptr) {
+      handle(val->definition());
+    } else {
+      setExact(val);
+    }
+  }
+
+  void handle(BinaryOp* bop) override {
+    const auto lhs = bop->lhs();
+    const auto rhs = bop->rhs();
+
+    handle(lhs);
+    handle(rhs);
+
+    if (!(isExact(lhs) && isExact(rhs))) {
+      return;
+    }
+
+    if (bop->getBinaryOpType() == BinaryOpType::CeilDiv) {
+      // CeilDiv is the only expression that can make an extent val
+      // larger than the actual. Need to know the exact values.
+      ExpressionEvaluator ee(fusion_);
+      const auto lhs_value = ee.evaluate(lhs);
+      const auto rhs_value = ee.evaluate(rhs);
+      if (lhs_value.has_value() && rhs_value.has_value() &&
+          (lhs_value.value() % rhs_value.value()) == 0) {
+        setExact(bop->out());
+      }
+    } else if (bop->getBinaryOpType() == BinaryOpType::Mul) {
+      setExact(bop->out());
+    } else {
+      // Expr on extent should be either CeilDiv or Mul, which are
+      // derived from split and merge, respectively.
+      TORCH_INTERNAL_ASSERT("Unexpected BinaryOpType: ", bop);
+    }
+  }
+
+ private:
+  Fusion* fusion_ = nullptr;
+  //! Vals that are known to need no predicate if used as IterDomain extent
+  std::unordered_set<const Val*> exact_vals_;
+};
+
+// TODO: Merge with IterationDomainAnalysis
+class PredicateAnalyzer : public OptOutDispatch {
+ public:
+  //! Checks if a predicate is needed to avoid out-of-bound accesses.
+  //!
+  //! Due to the way we allocate local-memory tensors, there should
+  //! never be out-of-bound accesses with consumer tensors when allocated on
+  //! local memory. However, accessing producer tensors still may
+  //! result in out-of-bound as they are replyaed as consumers.
```

Comment:
Is this from RFactor or reductions?

Reply:
I'm not sure I understand your question, but this isn't specific to Rfactor nor reductions.

Example:
```
t0: [I1]
t1: [I1]
t1 = t0 + 1

t1->split(0, 10);
t0->computeAt(t1, 1);
t1->split(1, 3);

t0: [I1/10, 10]
t1: [I1/10, 4, 3]
```

Suppose both `t0` and `t` are on registers. They are allocated as `[10]` and `[4 * 3]`, respectively. When executing `t1 = t0 + 1`, the loop nest would look like:

```
for (i = 0; i < 4; ++i) {
   for (j = 0; j < 3; ++j) {
     t1[i * 3 + j] = t0[i * 3 + j];
   }
}
```
In this case, `t1` is fine without predicates, but `t0` is not.

If the second split of `t1` is evenly divisible, the expression should be safe without predicates. That's what the code below tries to detect.